test/e2e/cert_rotation_test.go

```diff
@@ -0,0 +1,115 @@
+package e2e
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	appv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	// "sigs.k8s.io/cluster-api/test/e2e/internal/log"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/apimachinery/pkg/labels"
+	framework "sigs.k8s.io/cluster-api/test/framework"
+)
+
+func cert_rotation() {
+	Logf("Start the certificate rotation test")
+	By("Check if Ironic pod is ready")
+	targetClusterClientSet := targetCluster.GetClientSet()
+	ironicNamespace := os.Getenv("NAMEPREFIX") + "-system"
+	ironicDeploymentName := os.Getenv("NAMEPREFIX") + "-ironic"
+	ironicDeployment, err := getDeployment(targetCluster, ironicDeploymentName, ironicNamespace)
+	Eventually(func() error {
+		ironicPod, err := getPodFromDeployment(targetCluster, ironicDeployment, ironicNamespace)
+		if err != nil {
+			return err
+		}
+		if ironicPod.Status.Phase == corev1.PodRunning {
+			return nil
+		}
+
+		return errors.New("Ironic pod cannot be in running state")
+	}, e2eConfig.GetIntervals(specName, "wait-deployment")...).Should(BeNil())
+
+	By("Force the cert-manager to regenerate the certificate by deleting the secrets")
+
+	secretList := []string{
+		"ironic-cert",
+		"ironic-inspector-cert",
+		"mariadb-cert",
+	}
+	for _, secretName := range secretList {
+		err := targetClusterClientSet.CoreV1().Secrets(ironicNamespace).Delete(ctx, secretName, metav1.DeleteOptions{})
+		Expect(err).To(BeNil(), "Cannot detele the this secret: %s", secretName)
+	}
+
+	By("Wait for containers in the ironic pod to be restarted")
+	Logf("Get the current number of time containers were restarted")
+	containerNumRestart := make(map[string]int32)
+	containerNumRestart["ironic-api"] = 0
+	containerNumRestart["ironic-conductor"] = 0
+	containerNumRestart["mariadb"] = 0
+	Expect(err).To(BeNil())
+	ironicPod, err := getPodFromDeployment(targetCluster, ironicDeployment, ironicNamespace)
+	Expect(err).To(BeNil())
+	for _, container := range ironicPod.Status.ContainerStatuses {
+		if _, exist := containerNumRestart[container.Name]; exist {
+			containerNumRestart[container.Name] = container.RestartCount
+		}
+	}
+	Logf("Wait containers which got a new certificate until they are succesfully restarted")
+	Eventually(func() error {
+		ironicPod, err := getPodFromDeployment(targetCluster, ironicDeployment, ironicNamespace)
+		if err != nil {
+			return err
+		}
+
+		if ironicPod.Status.Phase == corev1.PodRunning {
+			for _, container := range ironicPod.Status.ContainerStatuses {
+				if oldNumRestart, exist := containerNumRestart[container.Name]; exist {
+					if !(oldNumRestart < container.RestartCount) {
+						return fmt.Errorf("%s is not restarted", container.Name)
```

Comment:
I am not sure if I am getting the logic properly here, would not this be always true? Also, can't pod restart in other cases than after regenerating certs, i.e node unavailability, resource limitation?

Reply:
I'm expecting that whenever the certificate is renewed, the container which is mounting it should restart itself to load the new cert into its process. That is true that something else may cause the restart. Do you have any ideas how we can improve it?

Reply:
I had an idea of comparing old and newly regenerated certs and once we are sure it is a new one, check if same pod restarted, but based on the conversation, I do agree that might not be a necessary thing to do, so I am fine leaving it as is.